[move-package-manifest] initial implementation

vgao1996 opened the pull request:
This PR introduces the first component of the next-generation package system: a new manifest parser.

## High level designs
- **Modular architecture**: The new package system is being developed as a collection of self-contained components, each intended to be reusable as a standalone library.
- **Single-responsibility**: Each component focuses on doing one thing well, with a clearly defined and narrow scope.
  - For this crate, that means defining the manifest AST and parsing logic only.
  - Package resolution is explicitly out of scope and will be handled separately.
- **Serde-based parsing**: Replaces the custom TOML parser with `serde`, making the codebase easier to maintain and less error-prone.

## Test coverage
The parser comes with a comprehensive test suite:
- **`expected_output_tests`**: Validates that the parser produces exactly the expected outputs using synthetic inputs.
- **`serde_round_trip`**: Verifies that manifests remain unchanged after a `serialize -> deserialize` round-trip using the same synthetic inputs.
- **`parse_framework_manifests`**: Ensures all manifests in `aptos-move/framework` can be successfully parsed.
- **`parse_move_example_manifests`**: Ensures all manifests in `aptos-move/move-examples` can be successfully parsed.

JakeSilverman commented on third_party/move/tools/move-package-manifest/src/manifest.rs:
```diff
@@ -0,0 +1,328 @@
+            version: Default::default(),
+            authors: BTreeSet::new(),
+            license: None,
+            upgrade_policy: None,
```
Do we want packages to ever have `none` as the upgrade_policy? Is `none` interpreted differently than `compatible`? And if not, shouldn't we just use `compatible` here as default?

vgao1996 replied:
`None` here allows us to express that the user has not specified an upgrade policy, which is subtly different from having compatible as a default. Being aware of this distinction may allow us to handle this better in later stages.

JakeSilverman commented on third_party/move/tools/move-package-manifest/src/manifest.rs:
```diff
@@ -0,0 +1,328 @@
+
+    /// List of authors.
+    #[serde(default)]
+    pub authors: BTreeSet<String>,
```
Should authors be a set or a list? If set, it is a bit counterintuitive that the author list ["alice", "alice"] doesn't give any kind of warning.

JakeSilverman commented on third_party/move/tools/move-package-manifest/src/manifest.rs:
```diff
@@ -0,0 +1,328 @@
+    where
+        D: Deserializer<'de>,
+    {
+        let make_err = || serde::de::Error::custom("Invalid version format");
```
the error message in `parse_version` of `move-package/src/source_package/manifest_parser.rs` seems much more robust error message for malformed version format. Do we want to keep this robust error message?

vgao1996 replied:
Good catch. I got the message improved.

JakeSilverman commented on third_party/move/tools/move-package-manifest/src/manifest.rs:
```diff
@@ -0,0 +1,328 @@
+    {
+        let raw = RawDependency::deserialize(deserializer)?;
+
+        let location = match (raw.local, raw.git, raw.aptos) {
```
This match means that expressions like

`a = { local = "../a", version = "1.3.3", rev = "main" }` are valid even though `rev` isn't really doing anything here. Same for `subdir`, `address`. So we want this?

vgao1996 replied:
Yeah was aware of this already but it wasn't the most straightforward to check. Finally came up with working solution though. 

It now errors on a field if associated tag name ("git", "aptos") is missing
```
error: failed to parse manifest
  |
1 | [package]
2 | name = "some_package_name"
3 | version = "0.1.2"
4 |
5 | [dependencies]
6 | a = { local = "../a", address = "0x123" }
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ redundant field "address" -- only needed for "aptos" dependencies
  |
```

JakeSilverman commented on third_party/move/tools/move-package-manifest/src/util.rs:
```diff
@@ -0,0 +1,53 @@
+            let snippet = Snippet::source(manifest_text)
+                .annotation(Level::Error.span(span).label(err.message()));
+
+            writeln!(
```
Nit: Can the manifest files ever be very large? In this case, it seems like printing out the entire manifest (as opposed to just the part with error) may be cumbersome for user.

vgao1996 replied:
The manifests are likely to remain reasonably small in the foreseeable future as we have a lot fewer options than Rust and Move projects likely have fewer dependencies, so I think it's fine to render the whole stuff for now.

Should future needs arise we can improve this easily.

vgao1996 commented:
@JakeSilverman thanks for the quality review! Appreciate that you caught many of these things I missed.

I've addressed all your comments. Please let me know if the PR looks good now!

vgao1996 commented on Cargo.lock:
```diff
@@ -18447,6 +18471,12 @@ version = "0.1.11"
 checksum = "e51733f11c9c4f72aa0c160008246859e340b00807569a0da0e7a1079b27ba85"
 
+[[package]]
+name = "unicode-width"
```
This is likely introduced by a dependency transitively. The lack of direct control makes it a bit harder for us to unify the two versions. IMO it's not a big issue, and if the semantic versioning is to be trusted, then the two versions are incompatible anyway.

vineethk replied:
The main concern is the compile times (compiling each package's each version once), and how these add up. I'll leave it to your judgement.

Note that if semantic versioning is to be trusted, then only the minor version & patch changed, so 0.2.0 is backwards compatible with 0.1.11.

vgao1996 replied:
> Note that if semantic versioning is to be trusted, then only the minor version & patch changed, so 0.2.0 is backwards compatible with 0.1.11.

Not an expert on this, but I recall there's a special rule about pre 1.0 versions, stating that anything could change
- https://semver.org/#spec-item-4

Again I'm not 100% sure, but if `0.2.0` is indeed considered compatible with `0.1.11`, then cargo would've unified the two. 

> The main concern is the compile times (compiling each package's each version once), and how these add up.

Valid concern, and a general issue for big Rust projects. It's tricky to fix though, cuz the two different versions are likely pulled in indirectly through transitive dependencies. There's not much you can do other than upgrading or downgrading your immediate dependencies and then hoping they'll end up having using the same version of a transitive dependency.

vgao1996 commented on third_party/move/tools/move-package-manifest/src/manifest.rs:
```diff
@@ -0,0 +1,392 @@
+
+/// Represents either a wildcard or a numerical address.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum AddressOrWildcard {
```
Renamed to `AddressAssignment`

vgao1996 commented on third_party/move/tools/move-package-manifest/src/manifest.rs:
```diff
@@ -0,0 +1,392 @@
+#[serde(deny_unknown_fields)]
+pub struct BuildOptions {
+    pub language_version: Option<Version>,
+}
```
Correct, but as we discussed earlier, it's yet to be seen whether we'll need a more powerful way to handle extensions in the future. For now this is good enough though.

vgao1996 commented on third_party/move/tools/move-package-manifest/tests/move_example_manifests.rs:
```diff
@@ -0,0 +1,34 @@
+    r"\bMove\.toml$"
+);
+
+fn parse_move_example_manifests(path: &Path) -> datatest_stable::Result<()> {
```
IMO the framework + move-examples already provide us with a pretty good testsuite. Going beyond them brings marginal benefits while adding a lot of coupling -- even the current setup could already turn out to be surprising to people working on the framework and example packages.

vineethk replied:
It might be worth doing this even one time, manually, and not commit it (but document the results in the PR description).

For a parser implementation looking to maintain backward compatibility (or find corner cases where you want to explicitly break it), more tests the better.

vgao1996 commented on third_party/move/tools/move-package-manifest/src/manifest.rs:
```diff
@@ -0,0 +1,392 @@
+    /// Refers to a package published on-chain.
+    // REVISIT: Potentially leaky abstraction -- do we still want this to be platform agnostic?
+    Aptos {
+        /// URL to the Aptos full-node connected to the network where the package is published.
```
So as I discussed with you, the design of the whole `Aptos` variant is still tentative, so the details aren't fixed it. I should be able to get these finalized as I explore the package resolver.

I've added a new TODO documenting this point and mentioning the design choices we'll need to make.

vgao1996 commented on third_party/move/tools/move-package-manifest/src/manifest.rs:
```diff
@@ -0,0 +1,392 @@
+        /// URL to the Aptos full-node connected to the network where the package is published.
+        node_url: String,
+
+        /// Address of the published package.
```
See previous comment.

vgao1996 commented on third_party/move/tools/move-package-manifest/src/manifest.rs:
```diff
@@ -0,0 +1,392 @@
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum PackageLocation {
+    /// Refers to a package stored in the local file system.
+    Local { path: String },
```
Switched to `PathBuf`.

> and the custom deserializer can even check if it is a valid path to a directory

No, I don't think this is the job of the deserializer -- it would be cleaner if it focuses on defining the AST and parsing it. 

Conceptually you can consider "checking if a path is valid" as an act of package resolution and it belongs to the next component -- the package resolver.

vgao1996 commented on third_party/move/tools/move-package-manifest/src/package_name.rs:
```diff
@@ -0,0 +1,107 @@
+
+/// A wrapper around a `String` representing the name of a Move package.
+///
+/// A valid package name must:
```
Correct. Documented this in a comment

```
/// TODO: The rules above are tentative and are subject to change if we find incompatibility
///       in production.
```

vgao1996 commented on aptos-move/move-examples/aggregator_examples/Move.toml:
```diff
@@ -1,5 +1,5 @@
 [package]
-name = "Aggregator examples"
+name = "Aggregator-examples"
```
Yeah, I'll need to survey all the production packages before we could finalize the rules. Will do that as a parallel task.

vgao1996 commented on third_party/move/tools/move-package-manifest/src/package_name.rs:
```diff
@@ -0,0 +1,107 @@
+    }
+}
+
+fn is_valid_package_name(s: &str) -> bool {
```
Hmm... the two have subtle differences -- hyphens are allowed in package names but not named addresses.

For now I'd prefer keeping these helpers in their local scope and not over-generalizing them.

vgao1996 commented on third_party/move/tools/move-package-manifest/src/manifest.rs:
```diff
@@ -0,0 +1,392 @@
+#[derive(Default, Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct BuildOptions {
+    pub language_version: Option<Version>,
```
Done. However, I had to add custom `Serialize` and `Deserialize` implementations for `LanguageVersion`, so that we can use numeric versions like "2.0" as opposed to variant names like "V2_0". Let me know if this is safe -- I hope we don't currently store serialized `LanguageVersion` anywhere else.

vgao1996 commented:
@vineethk Thanks for great review! Fixed a few things I missed with your help.

I've addressed all your comments. Let me know if the PR looks good now!

vineethk reviewed (approved):
A couple of comments to consider, but the PR looks good!

vineethk commented on third_party/move/tools/move-package-manifest/src/manifest.rs:
```diff
@@ -0,0 +1,421 @@
+    pub addresses: BTreeMap<NamedAddress, AddressAssignment>,
+
+    /// Dev-only named address bindings defined by the package.
+    /// Being dev-only means the bindings are only active in unit tests, now when being compiled regularly.
```
```suggestion
    /// Being dev-only means the bindings are only active in unit tests, not when being compiled regularly.
```
